[DO NOT MERGE][BEAM-8151] Create 1000 SdkWorker objects instead of the current limit of 12.

lukecwik opened the pull request:
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

lukecwik commented:
CC: @pabloem @TheNeuralBit @aaltay @robertwb @Hannah-Jiang

Follow-up for https://github.com/apache/beam/pull/9477

lukecwik commented:
CC: @acrites

acrites commented on sdks/python/apache_beam/runners/worker/sdk_worker.py:
```diff
@@ -269,6 +250,24 @@ def _monitor_process_bundle(self):
+    """Gets an idle worker or creates a new one if none are idle."""
+    try:
+      self.workers.get(wait=False)
+    except queue.Empty:
```
We're now creating a new worker any time we need work and the queue is empty. When we return the worker after processing, do we have a limit to the number allowed in the queue? If not, then the queue size will always grow to the high water mark of number needed.

lukecwik replied:
Yes, a future change can clean-up unused SDK workers.

lukecwik commented:
I ran the performance benchmark using the fn_api_runner_microbenchmark:

before (at commit 80e8fb59b6d785eb462e6a05b91d029fd8ff50c6):
```
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 1 of 10, per element time cost: 3.0014 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 2 of 10, per element time cost: 0.0394958 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 3 of 10, per element time cost: 0.0220158 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 4 of 10, per element time cost: 0.0160263 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 5 of 10, per element time cost: 0.0130143 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 6 of 10, per element time cost: 0.0111442 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 7 of 10, per element time cost: 0.00963751 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 8 of 10, per element time cost: 0.00856183 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 9 of 10, per element time cost: 0.00796899 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 10 of 10, per element time cost: 0.00760294 sec


Fixed cost   3.49717886575063
Per-element  0.003790947336139101
R^2          0.963500601942131
```

after:
```
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 1 of 10, per element time cost: 2.8741 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 2 of 10, per element time cost: 0.0401369 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 3 of 10, per element time cost: 0.0219585 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 4 of 10, per element time cost: 0.0161115 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 5 of 10, per element time cost: 0.0124523 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 6 of 10, per element time cost: 0.0108133 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 7 of 10, per element time cost: 0.00963711 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 8 of 10, per element time cost: 0.00862622 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 9 of 10, per element time cost: 0.00795245 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 10 of 10, per element time cost: 0.00736677 sec


Fixed cost   3.4638128309249883
Per-element  0.003727303158153187
R^2          0.9529920831060255
```

lukecwik commented:
I tried generating the workers on demand as can be seen in https://github.com/lukecwik/incubator-beam/commit/c743564d3dd029260c0d8a5e703856c9ddf2650a but I ran into lots of test issues which assumed that there was only a single SdkWorker processing tasks so this is a short term solution to fix a concurrency limit issue with the Python SDK.

lukecwik commented:
Updated perf numbers:
```
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 1 of 10, per element time cost: 2.8493 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 2 of 10, per element time cost: 0.0399395 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 3 of 10, per element time cost: 0.0252568 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 4 of 10, per element time cost: 0.0163703 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 5 of 10, per element time cost: 0.0129816 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 6 of 10, per element time cost: 0.0109879 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 7 of 10, per element time cost: 0.0098083 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 8 of 10, per element time cost: 0.00878269 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 9 of 10, per element time cost: 0.00803661 sec
run_single_pipeline, 1 element(s) at start, 100 growth per run: run 10 of 10, per element time cost: 0.00757915 sec


Fixed cost   3.6176186113357542
Per-element  0.0037115805365822525
R^2          0.9044127183436924
```

lukecwik commented:
No longer relevant due to #10123 being merged.
